### PR TITLE
Add curve and scheme to vk and proof

### PR DESCRIPTION
**Pull request by Schaeff**
Add two fields to the serialized proof and serialized verification key:
- curve: a string with the name of the curve used. Ex: "bn128"
- scheme: a string with the name of the proving scheme used. Ex: "g16"

Adjust deserialization to look at these fields first to then build the correct generic structs. This change is backward-compatible as the new formats are an extension of the old ones.

### CODE REVIEW

File: zokrates_js/src/lib.rs [hunk 1]
```diff
@@ -487,22 +496,51 @@ pub fn generate_proof(
...
+    let proof_scheme = SchemeParameter::try_from(
+        proof["scheme"]
+            .as_str()
+            .ok_or_else(|| JsValue::from_str("Invalid proof: missing field `scheme`"))?,
+    )
+    .map_err(|e| JsValue::from_str(&e))?;
+
+    if proof_curve != vk_curve {
+        return Err(JsValue::from_str(
+            "Proof and verification should have the same curve",
```
Comment (dark64) [inline zokrates_js/src/lib.rs]: `verification` -> `verification key`

File: zokrates_js/src/lib.rs [hunk 2]
```diff
@@ -487,22 +496,51 @@ pub fn generate_proof(
...
+        ));
+    }
+
+    if proof_scheme != vk_scheme {
+        return Err(JsValue::from_str(
+            "Proof and verification should have the same scheme",
```
Comment (dark64) [inline zokrates_js/src/lib.rs]: same as above

File: zokrates_cli/src/ops/export_verifier.rs [hunk 3]
```diff
@@ -32,67 +31,58 @@ pub fn subcommand() -> App<'static, 'static> {
...
 
-    // read vk file
-    let input_path = Path::new(sub_matches.value_of("input").unwrap());
-    let input_file = File::open(&input_path)
-        .map_err(|why| format!("Could not open {}: {}", input_path.display(), why))?;
-    let reader = BufReader::new(input_file);
-
-    let vk = serde_json::from_reader(reader)
-        .map_err(|why| format!("Could not deserialize verification key: {}", why))?;
+    let vk = serde_json::from_value(vk).map_err(|why| format!("{:?}", why))?;
```
Comment (dark64) [inline zokrates_cli/src/ops/export_verifier.rs]: is `{:?}` purposely used here?

File: zokrates_cli/src/ops/verify.rs [hunk 4]
```diff
@@ -19,142 +19,177 @@ use zokrates_field::{Bls12_377Field, Bls12_381Field, Bn128Field, Bw6_761Field, F
...
+        .ok_or_else(|| "`curve` should be a string".to_string())?;
+    let vk_scheme = vk
+        .get("scheme")
+        .ok_or_else(|| "Field `scheme` not found in verification key".to_string())?
+        .as_str()
+        .ok_or_else(|| "`scheme` should be a string".to_string())?;
+
+    if proof_curve != vk_curve {
+        return Err(format!(
+            "Expected the curve of the proof and the verification to be equal, found {} != {}",
```
Comment (dark64) [inline zokrates_cli/src/ops/verify.rs]: `verification` -> `verification key`

File: zokrates_cli/src/ops/verify.rs [hunk 5]
```diff
@@ -19,142 +19,177 @@ use zokrates_field::{Bls12_377Field, Bls12_381Field, Bn128Field, Bw6_761Field, F
...
+            proof_curve, vk_curve
+        ));
+    }
+
+    if proof_curve != vk_curve {
+        return Err(format!(
+            "Expected the scheme of the proof and the verification to be equal, found {} != {}",
```
Comment (dark64) [inline zokrates_cli/src/ops/verify.rs]: same as above

File: zokrates_js/metadata.json [hunk 6]
```diff
@@ -0,0 +1 @@
+{"version":"0.7.13"}
```
Comment (dark64) [inline zokrates_js/metadata.json]: remove this file

File: zokrates_js/src/lib.rs [hunk 7]
```diff
@@ -487,22 +496,51 @@ pub fn generate_proof(
...
+        ));
+    }
+
+    let scheme = vk_scheme;
+    let curve = vk_curve;
```
Comment (dark64) [inline zokrates_js/src/lib.rs:542]: What's the reason for this assignment?

File: zokrates_js/stdlib.json [hunk 8]
```diff
@@ -0,0 +1 @@
...9137008396665083714591338741616893578930275511205,\n      5027542104048754930085470328670427788489455916338375169351586496298129661248,\n      1922685817237874482932428650501872692326329693528175054457715565489676406535,\n      3071473720617798005831658342971536643616129392641449174655528578463370685788,\n      21091078808046042460442535848913779439792606439995062001271357804782672390627,\n      19773167374024045118471391738750949555178717045037157435777574972149053404157,\n      6418695831178793575992210834992785624340084513619644969535805236049937971859,\n      6317875495482489567338519005308431806047606843913867465201005132273298011425,\n      18001249545956637376455848019549801116909661454019565655561439372098476761813,\n      15530167556609139699164228289904946047951254183080358784988008899829027775935,\n      8702757129830652230304011519426558036441096750485189115358314568895250616455,\n      6369986882953061252605652398893489899416599935424066958291402945530517772170,\n      6842894437627604179732847187262933342846269043996061072487488027804029200046,\n      20951621154051947571647917571547811655800779287153833018533872651413529893817,\n      12192775350807491348052917259375
```
Comment (dark64) [inline zokrates_js/stdlib.json]: remove this file

File: zokrates_js/src/lib.rs (hunk 7 above)
Comment (Schaeff) [reply zokrates_js/src/lib.rs:542]: Just clarity for the rest of the code

**Review by dark64 — APPROVED** [review]
LGTM

**Pull request merged**